### DISCUSSION TITLE
exploring-the-dotnet-8-preview-keyed-services-dependency-injection-support/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#385] [Announcements]
# exploring-the-dotnet-8-preview-keyed-services-dependency-injection-support/

In this post I discuss the new "keyed service" support for the dependency injection container, introduced in .NET 8 preview 7

https://andrewlock.net/exploring-the-dotnet-8-preview-keyed-services-dependency-injection-support/

**Comment by jakenuts** [comment]
For a cheap knockoff of this I define a generic interface for the service using a set of classes/records as the type parameter and register them like IKeyedDynamoDb<Production> and then use those in the constructors that want a particular service. Also several KeyedService nuget packages which approximate this feature if 8.x isn’t in your immediate future.

**Reply by jakenuts** [reply]
Oops, it gobbled the parameter, IKeyedDynamoDb[Production]

**Comment by kwaazaar** [comment]
Since this is all part of nuget packages (Microsoft.Extensions.DependencyInjection.*), will be be able to use this is older .NET versions as well?

Support for named services can be helpful when migrating code currently using one of the other DI frameworks. Even when you don't like named services, you can first migrate and later 'optimize' them away.

**Comment by gao-artur** [comment]
The old `Try` extensions returning `void` instead of `IServiceCollection` and it was admitted to be a mistake that will never be fixed to not break binary compatibility. I don't understand why the new keyd extensions are `void`. I don't think consistency should win the conviniece.

**Comment by wangengzheng** [comment]
good vote

**Comment by ShirazAdam** [comment]
This is nothing new. We’ve been able to do this using .NET Framework with Prism/Unity over a decade ago. Just Microsoft releasing the same thing with a new coat of paint. The only real difference is that .NET 8 is cross-platform.

**Comment by luizfbicalho** [comment]
For me is missing one api to get all the keys for one service

**Comment by kicsiede** [comment]
2023, ladies and gentlemen

**Comment by SViradiya-MarutiTech** [comment]
There is support in Minimal API now https://github.com/dotnet/aspnetcore/pull/50095, So I think you should update this document

**Reply by andrewlock** [reply]
As I explain in the post, this post was about the previews, and some features may have changed 🙂

**Comment by 0xfeeddeadbeef** [comment]
One would naively expect that keyed services work in conjunction with named IOptions, but this is not the case. Named IOptions is a completely independent thing.

**Comment by lafar6502** [comment]
Hi, just doing some tests of Microsoft DI container
And noticed one thing - when you register a keyed service, it won't be resolved with the keyless GetService version.
Only if the key is null the keyless GetService will get the instance.
So you cannot register several keyed instances and iterate over them with GetServices() - the keyed ones will not be resolved/returned in the resulting collection.

```
var services = new ServiceCollection();
            services.AddKeyedSingleton<ITestMe, T1>("K1");
            services.AddKeyedSingleton<ITestMe, T1>("K2");
            services.AddKeyedSingleton<ITestMe, T2>("K3");

            
            //services.AddKeyedSingleton<ITestMe, T2>(null);

            var container = services.BuildServiceProvider();

            
            var s0 = container.GetService<ITestMe>(); //NULL

            var s1 = container.GetKeyedService<ITestMe>("K1");
            var s2 = container.GetKeyedService<ITestMe>("K2");
            var s3 = container.GetKeyedService<ITestMe>("K3");

            var sss = container.GetServices<ITestMe>(); ///EMPTY collection returned
```

**Comment by maxisam** [comment]
i don't think you can inject IKeyedServiceProvider

https://github.com/dotnet/runtime/issues/90528#issuecomment-1680759846
